### DISCUSSION TITLE
Patterns and libraries for writing CLIs in Nushell

### DISCUSSION THREAD

**Discussion by chriscalo** [nushell/nushell#11969] [Q&A] [answered]
I'm considering rewriting some Node.js CLIs in Nushell. In Node.js I use a framework called [commander][] to handle the generic aspects of writing CLIs such as:
- declaratively describing commands, arguments, and options
- parsing user input and mapping to the correct subcommands
- auto-generating help output

[commander]: https://www.npmjs.com/package/commander

Here's an example of the programming model:

```js
const { Command } = require('commander');
const program = new Command();

program
  .name('string-util')
  .description('CLI to some JavaScript string utilities')
  .version('0.8.0');

program.command('split')
  .description('Split a string into substrings and display as an array')
  .argument('<string>', 'string to split')
  .option('--first', 'display just the first substring')
  .option('-s, --separator <char>', 'separator character', ',')
  .action((str, options) => {
    const limit = options.first ? 1 : undefined;
    console.log(str.split(options.separator, limit));
  });

program.parse();
```

And here's an example of the auto-generated help output:

```repl
$ node string-util.js help split
Usage: string-util split [options] <string>

Split a string into substrings and display as an array.

Arguments:
  string                  string to split

Options:
  --first                 display just the first substring
  -s, --separator <char>  separator character (default: ",")
  -h, --help              display help for command

$ node string-util.js split --separator=/ a/b/c
[ 'a', 'b', 'c' ]
```

# Question

I'm loving Nushell and am considering moving some CLIs from Node.js to Nushell. But it's probably going to be painful without utilities that handle some of these aspects.

**What are some patterns and/or libraries for handling these same considerations when writing Nushell CLIs?**

**Comment by ghost** [comment, marked as answer]
you don't need a library for these things (if i understood the question correct):

```nu
# list …
def "main list" [
  --all(-a)  # include …
]: nothing -> table {
  # code
}

def "complete_ids" [] {
  [1 2 3]  # (just return a list)
}

# kill ones with indexes
def "main kill" [
  --force(-f)
  ...id: int@complete_ids
] {
  # code
}
```

will generate the command `FILENAME list` and `FILENAME kill` with:
- documentation (the comment above, the arguments, the argument comment, the types, etc)
- parsing of subcommands (list and kill) including type validation, etc
- custom tab completion

i haven't published it yet, but i've been working on libraries and tbh except for some things like `[a b c] | index-of b` and progress-bars its pretty empty since most things i want are already there

**Reply by chriscalo** [reply]
Ah, you're right! I see that now, thanks so much for the reply.

After some exploration and investigation, I learned two useful things:

1. The [custom commands](https://www.nushell.sh/book/custom_commands.html) page has a section on [Documenting your command](https://www.nushell.sh/book/custom_commands.html#documenting-your-command) explaining how to use types and comments to document custom commands.

2. As far as I can tell, you can't run `help myscript.nu`, but it seems `./myscript.nu -h` DOES work, so that's how I was able to test the help output.

Thanks again.

**Comment by chriscalo** [comment]
After converting a script from Node.js with `commander` to Nushell, here are some of the non-obvious parts I ran into and some of the patterns I ended up with. I expect there are better ways to do some of these things, which is half the reason I'm sharing them here. (The other half being in case they're useful to anyone else writing CLIs in Nushell for the first time.)

# Pattern 1: `not-empty` variables

I found myself frequently ensuring that commands return a non-empty result. Puzzlingly, the language only provides an `is-empty` filter command. So I defined an opposite `not-empty` filter command:

```nu
def "not-empty" [] {
  not ($in | is-empty)
}
```

A usage might look something like:

```nu
if (git diff | str trim | not-empty) {
  error make {
    msg: "You have uncommitted changes. Commit before deploying."
  }
}
```

# Pattern 2: conditional parameters

When building a command to run, I found myself needing to conditionally set parameters based on user input. I ended up building up a list of parameters and joining them on spaces into a single string.

```nu
mut args = [
  "gcloud app deploy app.yaml"
  "--no-promote"
  $"--project=($project_id)"
  $"--version=($version_name)"
]

if ($quiet) {
  $args = ($args | append [ "--quiet" ])
}

let cmd = ($args | str join " ")
```

Honestly, I wish there were a cleaner, more concise way to do this than variable reassignment, but I wasn't aware of any such way.

# Pattern 3: executing a string as a command

When the command you want to execute is mostly or entirely dynamically built, it seems the most efficient way to run it is to store the command in a string variable and invoke it with `nu -c`:

```nu
let cmd = ($args | str join " ")

nu -c $cmd
```

# Pattern 4: temporary `$env` overrides

It's common to invoke a command with modified values for specific environment variables. `with-env` provides a clean and easy way to override specific environment variables but only for the life of the block that it's attached to.

```nu
with-env [
  # set the working directory
  PWD $project_root
] {
  nu -c $cmd
}
```

# Pattern 5: crawl up the file system hierarchy looking for a file

I needed to locate a config file without hard-coding its path. The most sensible way to do this seemed to be crawling up the file system hierarchy until the file was located. This one was a pain to get right, so if there's a simpler way to do this, I love to hear about it.

```nu
def "find-project-root" [] {
  mut $current_dir = $env.FILE_PWD
  
  while true {
    # Check if the file exists in the current directory
    let contains_app_yaml = (ls $current_dir)
      | where ($it.name | str ends-with "/app.yaml")
      | not-empty
    
    if ($contains_app_yaml) {
      return $current_dir
    } else if ($current_dir == "/") {
      error make {
        msg: $"app.yaml not found in parent directories of ($env.FILE_PWD)"
      }
    } else {
      # continue up a level
      $current_dir = ($current_dir | path dirname)
    }
  }
}
```

**Reply by ghost** [reply]
a few of my workarounds for these problems:

> Pattern 3: executing a string as a command

you can pass a list as arguments to a command, which is usually more conventient (especially with your example from 2):
```nu
git ...["status" "--short"]
let args = ["commit" "-m" "tmp"]
git ...$args
```
> Pattern 2: conditional parameters

for conditional parameters i usually put a `if` directly into the array creation:
```nu
git ...[
  "status"
  (if $short {"--short"})
]
```

> Pattern 4: temporary $env overrides

just like in bash if its just for one command you can prepend them:
```nu
PWD="/tmp/foo" HOME="/tmp/fakehome" CC="tcc" make
```

also env variables are already scoped to blocks:
```
echo $env.PWD  # /home/user
ls | each {|i|
  cd $i.name
  echo $env.PWD  # /home/Downloads …
  $env.FOO = 1
}
echo $env.PWD  # /home/user
echo $env.FOO  # Error: nu::shell::column_not_found
```

> Pattern 5: crawl up the file system hierarchy looking for a file

i use this (not pretty either, but its shorter):
```nu
while $env.PWD != '/' {
        cd ..
        if (".git" | path exists) { return $env.PWD }
}
null
```

**Reply by chriscalo** [reply]
These are much simpler approaches than what I was doing. I learned a lot from this so thank you again. I'm going go apply some of these more elegant patterns. 👍🏼

**Comment by fdncred** [comment]
Maybe this will land and make checking for not empty easier https://github.com/nushell/nushell/pull/11991

**Reply by chriscalo** [reply]
Happy to see this is under consideration 👍🏼
